### PR TITLE
client:{dcr,eth}: Live wallet reconfiguration

### LINKED ISSUE #1727: Implement LiveReconfigurer on dcr and eth
... and consider making it a required API for `asset.Wallet`.

Following up #1686

### PR DESCRIPTION
**Pull request by martonp**
Closes #1727

### CODE REVIEW

File: client/asset/btc/btc.go [hunk 1]
```diff
@@ -486,8 +486,8 @@ type WalletConfig struct {
 	FallbackFeeRate  float64 `ini:"fallbackfee"`
 	FeeRateLimit     float64 `ini:"feeratelimit"`
 	RedeemConfTarget uint64  `ini:"redeemconftarget"`
-	ActivelyUsed     bool    `ini:"special:activelyUsed"` // injected by core
-	Birthday         uint64  `ini:"walletbirthday"`       // SPV
+	ActivelyUsed     bool    `ini:"specialactivelyUsed"` // injected by core
```
Comment (martonp) [inline client/asset/btc/btc.go:489]: The colon no longer works in the ini package for some reason:
https://github.com/decred/dcrdex/commit/2b4ce79c61302044a045b3fb5a85fcf6b71fa2e6

File: client/asset/eth/eth.go [hunk 2]
```diff
@@ -504,6 +506,23 @@ func (w *TokenWallet) Connect(ctx context.Context) (*sync.WaitGroup, error) {
...
 }
 
+// Reconfigure attempts to reconfigure the wallet.
+func (w *ETHWallet) Reconfigure(ctx context.Context, cfg *asset.WalletConfig, currentAddress string) (restart bool, err error) {
+	walletCfg, err := parseWalletConfig(cfg.Settings)
+	if err != nil {
+		return false, err
+	}
+
+	w.baseWallet.gasFeeLimit = walletCfg.GasFeeLimit
```
Comment (chappjc) [inline client/asset/eth/eth.go:516]: Looks like `gasFeeLimit` needs to be an atomic field.

File: client/cmd/dexc/config.go [hunk 3]
```diff
@@ -23,7 +23,7 @@ const (
 	defaultRPCKeyFile  = "rpc.key"
 	defaultMainnetHost = "127.0.0.1"
 	defaultTestnetHost = "127.0.0.2"
-	defaultSimnetHost  = "127.0.0.3"
+	defaultSimnetHost  = "127.0.0.1"
```
Comment (chappjc) [inline client/cmd/dexc/config.go:26]: This shouldn't change.

**Review by JoeGruffins — APPROVED** [review]
Looks good and working well for me.

I notice there is the `load from file` button on eth which has no function currently. It is also there when looking at the btc and dcr built-in wallets. Is it possible to hide it whenever one is looking at an spv/built-in wallet form?

image

File: client/asset/dcr/dcr.go [hunk 4]
```diff
@@ -808,23 +828,57 @@ func (dcr *ExchangeWallet) Connect(ctx context.Context) (*sync.WaitGroup, error)
...
 }
 
+// Reconfigure attempts to reconfigure the wallet.
+func (dcr *ExchangeWallet) Reconfigure(ctx context.Context, cfg *asset.WalletConfig, currentAddress string) (restart bool, err error) {
+	dcrCfg := new(walletConfig)
+	_, err = loadConfig(cfg.Settings, dcr.network, dcrCfg)
+	if err != nil {
+		return false, err
+	}
+
```
Comment (buck54321) [inline client/asset/dcr/dcr.go:838]: In `NewWallet`, we're doing 
```Go
// dcrwallet doesn't accept an empty string as an account name. Use the
// default value.
if walletCfg.PrimaryAccount == "" {
	walletCfg.PrimaryAccount = defaultAcctName
}
```

I had used this feature when creating my Decred wallet, and so when I went to reconfigure, I got an error `-11: account not found`.

**Review by buck54321 — CHANGES_REQUESTED** [review]

File: client/asset/btc/btc.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (martonp) [reply client/asset/btc/btc.go:489]: Underscore works.

File: client/asset/dcr/dcr.go [hunk 5]
```diff
@@ -506,29 +506,36 @@ type feeStamped struct {
...
-	useSplitTx       bool
 	apiFeeFallback   bool
+}
+
+// ExchangeWallet is a wallet backend for Decred. The backend is how the DEX
+// client app communicates with the Decred blockchain and wallet. ExchangeWallet
+// satisfies the dex.Wallet interface.
+type ExchangeWallet struct {
+	cfgV atomic.Value // *exchangeWalletConfig
+
```
[the remark this answers is not in the document]
Comment (martonp) [reply client/asset/dcr/dcr.go:530]: What do you mean by alignment bug?

[the remark this answers is not in the document]
Comment (chappjc) [reply client/asset/dcr/dcr.go:530]: https://pkg.go.dev/sync/atomic#pkg-note-BUG

When we have a have a 64-bit value that we intend to access with the sync/atomic functions, we have to do some tricks to make sure it's 64-bit aligned on 32-bit platforms.

Not an issue with `atomic.Value`, so it doesn't need to be at the top or anywhere else in particular.

File: client/asset/dcr/dcr_test.go [hunk 6]
```diff
@@ -2611,3 +2622,99 @@ func Test_dcrPerKBToAtomsPerByte(t *testing.T) {
...
+		t.Fatalf("failed to mapify: %v", err)
+	}
+
+	walletCfg := &asset.WalletConfig{
+		Type:     walletTypeDcrwRPC,
+		Settings: settings1,
+		DataDir:  "abcd",
+	}
+
+	reconfigurer.restart = false
```
[the remark this answers is not in the document]
Comment (buck54321) [reply client/asset/dcr/dcr_test.go]: I meant we should test with `reconfigurer.restart = true`.

**Review by buck54321 — APPROVED** [review]
A couple optional nits from previous review, but nothing to hold up approval. Nice work!

File: client/asset/dcr/rpcwallet.go [hunk 7]
```diff
@@ -190,6 +191,72 @@ func newRPCWallet(settings map[string]string, logger dex.Logger, net dex.Network
...
+		a, err := stdaddr.DecodeAddress(currentAddress, w.chainParams)
+		if err != nil {
+			return false, err
+		}
+		owns, err := newWallet.AccountOwnsAddress(ctx, a, depositAccount)
+		if err != nil {
+			return false, err
+		}
+		if !owns {
+			return false, errors.New("cannot reconfigure to different wallet while there are active deals")
```
Comment (chappjc) [inline client/asset/dcr/rpcwallet.go:245]: I know what you're saying, but "deals" doesn't jive with all the rest of our messages

**Review by chappjc — APPROVED** [review]
All working well for me.  The "deals" wording deviates to the point of being potentially confusing, but otherwise all looks good.

File: client/asset/dcr/rpcwallet.go (hunk 7 above)
Comment (martonp) [reply client/asset/dcr/rpcwallet.go]: changed it to "trades".

**Pull request merged**